modules/friendlytag.js

```diff
@@ -1222,69 +1544,159 @@ Twinkle.tag.callbacks = {
+				if(params.altLangFrom) {
+					tagText += '|from=' + params.altLangFrom;
+				}
+				if(params.altLangTo) {
+					tagText += '|to=' + params.altLangTo;
 				}
 			}
+			tagText += '}}';
+
+			if ( tagIndex > 0 ) {
+				if( tagIndex === (tags.length - 1) ) {
+					summaryText += ' and';
+				} else if ( tagIndex < (tags.length - 1) ) {
+					summaryText += ',';
+				}
+			}
+
+			summaryText += ' {{[[:' + (tagName.indexOf(":") !== -1 ? tagName : ("Template:" + tagName + "|" + tagName)) + ']]}}';
 		}
 
 		tags.sort();
-		totalTags = tags.length;
 		$.each(tags, addTag);
 
-		if( Twinkle.tag.mode === 'redirect' ) {
-			// Check for all Rcat shell redirects (from #433)
-			if (pageText.match(/{{(?:redr|this is a redirect|r(?:edirect)?(?:.?cat.*)?[ _]?sh)/i)) {
-				// Regex courtesy [[User:Kephir/gadgets/sagittarius.js]] at [[Special:PermaLink/831402893]]
-				var oldTags = pageText.match(/(\s*{{[A-Za-z ]+\|)((?:[^|{}]*|{{[^|}]*}})+)(}})\s*/i);
-				pageText = pageText.replace(oldTags[0], oldTags[1]+tagText+oldTags[2]+oldTags[3]);
-			} else {
-				// Fold any pre-existing Rcats into taglist and under Rcatshell
-				var pageTags = pageText.match(/\n{{R(?:edirect)? .*?}}/img);
-				var oldPageTags ='';
-				if (pageTags) {
-					pageTags.forEach(function(pageTag) {
-						var pageRe = new RegExp(pageTag, 'img');
-						pageText = pageText.replace(pageRe,'');
-						oldPageTags = oldPageTags.concat(pageTag);
-					});
-				}
-				pageText += '\n{{Redirect category shell|' + tagText + oldPageTags + '\n}}';
-			}
+		// Check for all Rcat shell redirects (from #433)
+		if (pageText.match(/{{(?:redr|this is a redirect|r(?:edirect)?(?:.?cat.*)?[ _]?sh)/i)) {
+			// Regex courtesy [[User:Kephir/gadgets/sagittarius.js]] at [[Special:PermaLink/831402893]]
+			var oldTags = pageText.match(/(\s*{{[A-Za-z ]+\|)((?:[^|{}]*|{{[^|}]*}})+)(}})\s*/i);
+			pageText = pageText.replace(oldTags[0], oldTags[1] + tagText + oldTags[2] + oldTags[3]);
 		} else {
-			// Smartly insert the new tags after any csd or prod templates
-			// or hatnotes; afd not yet supported since it places a comment
-			// not a template first. Regex is more complicated than it needs to be,
-			// which allows templates as parameters and to handle whitespace properly.
-			pageText = pageText.replace(/^\s*(?:((?:\s*\{\{\s*(?:db|delete|db-.*?|speedy deletion-.*?|(?:proposed deletion|prod blp)\/dated|about|correct title|dablink|distinguish|for|other\s?(?:hurricaneuses|people|persons|places|uses(?:of)?)|redirect(?:-acronym)?|see\s?(?:also|wiktionary)|selfref|the)\d*\s*(\|(?:\{\{[^{}]*\}\}|[^{}])*)?\}\})+(?:\s*\n)?)\s*)?/i,
-				"$1" + tagText);
+			// Fold any pre-existing Rcats into taglist and under Rcatshell
+			var pageTags = pageText.match(/\n{{R(?:edirect)? .*?}}/img);
+			var oldPageTags ='';
+			if (pageTags) {
+				pageTags.forEach(function(pageTag) {
+					var pageRe = new RegExp(pageTag, 'img');
+					pageText = pageText.replace(pageRe,'');
+					oldPageTags.push(pageTag);
```

Comment:
Right you are! I blindly replaced all `x = x.concat(y)` with `x.push(y)`, not realising `concat` is available for strings too. Since this is a string, I guess `x += y` should be clearer.